Added a test that reproduces this on 2.0 branch typically within 2 seconds, but does not manifest against this branch.  If it fails, you should see something like:

```
    2743   [cluster1-nio-worker-3] ERROR com.datastax.driver.core.ReusedStreamIdTest  - Got response without column schema_version, got columns Columns[thrift_version(varchar)] from Host /127.0.1.2:9042.
    2743   [cluster1-nio-worker-2] ERROR com.datastax.driver.core.ReusedStreamIdTest  - Got response without column rack, got columns Columns[release_version(varchar)] from Host /127.0.1.1:9042.
    2744   [cluster1-nio-worker-1] ERROR com.datastax.driver.core.ReusedStreamIdTest  - Got response without column thrift_version, got columns Columns[schema_version(uuid)] from Host /127.0.1.1:9042.
    3749   [cluster1-nio-worker-2] ERROR com.datastax.driver.core.ReusedStreamIdTest  - Got response without column broadcast_address, got columns Columns[rack(varchar)] from Host 

java.lang.AssertionError: Error triggered at or before 20 of 500 requests after 2040ms.
```

--- a/changelog/README.md
+++ b/changelog/README.md
@@ -1,5 +1,10 @@
 ## Changelog
 
+### 2.0.12.2 (in progress)
+
+- [bug] JAVA-1179: Request objects should be copied when executed.
+
+
 ### 2.0.12.1
 
 - [bug] JAVA-994: Don't call on(Up|Down|Add|Remove) methods if Cluster is closed/closing.

--- a/driver-core/src/main/java/com/datastax/driver/core/Connection.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/Connection.java
@@ -502,11 +502,10 @@ class Connection {
 
     public ResponseHandler write(ResponseCallback callback, boolean startTimeout) throws ConnectionException, BusyConnectionException {
 
-        Message.Request request = callback.request();
-
         ResponseHandler handler = new ResponseHandler(this, callback);
         dispatcher.add(handler);
-        request.setStreamId(handler.streamId);
+
+        Message.Request request = callback.request().setStreamId(handler.streamId);
 
         /*
          * We check for close/defunct *after* having set the handler because closing/defuncting

--- a/driver-core/src/main/java/com/datastax/driver/core/Message.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/Message.java
@@ -46,7 +46,7 @@ abstract class Message {
         public R decode(ByteBuf body);
     }
 
-    private volatile int streamId;
+    private volatile int streamId = -1;
 
     protected Message() {
     }
@@ -101,6 +101,19 @@ abstract class Message {
             this.tracingRequested = tracingRequested;
         }
 
+        @Override
+        public Request setStreamId(int streamId) {
+            // JAVA-1179: defensively guard against reusing the same Request object twice.
+            // If no streamId was ever set we can use this object directly, otherwise make a copy.
+            if (getStreamId() < 0)
+                return (Request) super.setStreamId(streamId);
+            else {
+                Request copy = this.copy();
+                copy.setStreamId(streamId);
+                return copy;
+            }
+        }
+
         public boolean isTracingRequested() {
             return tracingRequested;
         }
@@ -140,9 +153,7 @@ abstract class Message {
             }
         }
 
-        Request copy() {
-            throw new UnsupportedOperationException();
-        }
+        abstract Request copy();
 
         Request copy(ConsistencyLevel newConsistencyLevel) {
             throw new UnsupportedOperationException();

--- a/driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java
@@ -64,7 +64,7 @@ class RequestHandler {
     private final long startTime;
 
     private final AtomicBoolean isDone = new AtomicBoolean();
-    private AtomicInteger executionCount = new AtomicInteger();
+    private final AtomicInteger executionCount = new AtomicInteger();
 
     public RequestHandler(SessionManager manager, Callback callback, Statement statement) {
         this.id = Long.toString(System.identityHashCode(this));
@@ -106,10 +106,6 @@ class RequestHandler {
 
         Message.Request request = callback.request();
         int position = executionCount.incrementAndGet();
-        // Clone the request after the first execution, since we set the streamId on it later and we
-        // don't want to share that across executions.
-        if (position > 1)
-            request = request.copy();
 
         SpeculativeExecution execution = new SpeculativeExecution(request, position);
         runningExecutions.add(execution);

--- a/driver-core/src/main/java/com/datastax/driver/core/Requests.java
+++ b/driver-core/src/main/java/com/datastax/driver/core/Requests.java
@@ -46,15 +46,22 @@ class Requests {
         };
 
         private final Map<String, String> options;
+        private final ProtocolOptions.Compression compression;
 
         public Startup(ProtocolOptions.Compression compression) {
             super(Message.Request.Type.STARTUP);
+            this.compression = compression;
 
             ImmutableMap.Builder<String, String> map = new ImmutableMap.Builder<String, String>();
             map.put(CQL_VERSION_OPTION, CQL_VERSION);
             if (compression != ProtocolOptions.Compression.NONE)
                 map.put(COMPRESSION_OPTION, compression.toString());
             this.options = map.build();
+        }
+
+        @Override
+        Request copy() {
+            return new Startup(compression);
         }
 
         @Override
@@ -83,6 +90,11 @@ class Requests {
             super(Message.Request.Type.CREDENTIALS);
             this.credentials = credentials;
         }
+
+        @Override
+        Request copy() {
+            return new Credentials(credentials);
+        }
     }
 
     public static class Options extends Message.Request {
@@ -98,6 +110,11 @@ class Requests {
 
         public Options() {
             super(Message.Request.Type.OPTIONS);
+        }
+
+        @Override
+        Request copy() {
+            return new Options();
         }
 
         @Override
@@ -432,6 +449,11 @@ class Requests {
         }
 
         @Override
+        Request copy() {
+            return new Prepare(query);
+        }
+
+        @Override
         public String toString() {
             return "PREPARE " + query;
         }
@@ -462,6 +484,11 @@ class Requests {
         }
 
         @Override
+        Request copy() {
+            return new Register(eventTypes);
+        }
+
+        @Override
         public String toString() {
             return "REGISTER " + eventTypes;
         }
@@ -485,6 +512,11 @@ class Requests {
         public AuthResponse(byte[] token) {
             super(Message.Request.Type.AUTH_RESPONSE);
             this.token = token;
+        }
+
+        @Override
+        Request copy() {
+            return new AuthResponse(token);
         }
     }
 }

--- a/driver-core/src/test/java/com/datastax/driver/core/ReusedStreamIdTest.java
+++ b/driver-core/src/test/java/com/datastax/driver/core/ReusedStreamIdTest.java
@@ -1,0 +1,171 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import com.datastax.driver.core.exceptions.NoHostAvailableException;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.Uninterruptibles;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.testng.Assert.fail;
+
+public class ReusedStreamIdTest {
+
+    static Logger logger = LoggerFactory.getLogger(ReusedStreamIdTest.class);
+
+    private CCMBridge ccm;
+
+    @BeforeClass(groups = "long")
+    public void setupCluster() throws InterruptedException {
+        ccm = CCMBridge.builder("test")
+                .withNodes(2)
+                .build();
+    }
+
+    @AfterClass(groups = "long")
+    public void shutdownCluster() {
+        if (ccm != null)
+            ccm.stop();
+    }
+
+    /**
+     * Ensures that if activity tied to future completion blocks netty io threads that this does not cause the
+     * driver to possibly invoke the wrong handler for a response, as described in JAVA-1179.
+     * <p>
+     * This is accomplished by setting up a 2 node cluster and setting a low read timeout (1 second).  The test
+     * submits 10 concurrent requests repeatedly for up to 500 queries and on completion of each request may block
+     * in a callback randomly (25% of the time) for 1 second, causing a retry on the next host to trigger.  If a new
+     * stream id is not allocated on the retry, its possible it could use an already used stream id and cause the driver
+     * to invoke the wrong handlers for a response.   The test checks for this by ensuring that the column returned in a
+     * response matches the one queried.  If the column received does not match, the test fails.  In cases where this
+     * bug is present, it should be detected within 10 seconds.
+     *
+     * @jira_ticket JAVA-1179
+     * @test_category queries:async_callback
+     */
+    @Test(groups = "long")
+    public void should_not_receive_wrong_response_when_callbacks_block_io_thread() {
+        // Low Read Timeout to trigger retry behavior.
+        Cluster cluster = Cluster.builder()
+                .addContactPoint(CCMBridge.IP_PREFIX + '1')
+                .withSocketOptions(new SocketOptions().setReadTimeoutMillis(1000))
+                .build();
+
+        int concurrency = 10;
+        final Semaphore semaphore = new Semaphore(concurrency);
+        // RNG to determine sleep times.
+        final Random random = new Random();
+
+        try {
+            Session session = cluster.connect();
+
+            // Use the system.local table and alternate between columns that are queried.
+            List<ColumnMetadata> columnsToGrab = cluster.getMetadata().getKeyspace("system").getTable("local").getColumns();
+            assertThat(columnsToGrab.size()).isGreaterThan(1);
+
+            final CountDownLatch errorTrigger = new CountDownLatch(1);
+
+            long start = System.currentTimeMillis();
+            // 500 iterations will take roughly 1 minute.
+            int iterations = 500;
+            final AtomicInteger completed = new AtomicInteger(0);
+
+            for (int i = 1; i <= iterations; i++) {
+                try {
+                    if (errorTrigger.getCount() == 0) {
+                        fail(String.format("Error triggered at or before %d of %d requests after %dms.", i, iterations,
+                                System.currentTimeMillis() - start));
+                    }
+                    semaphore.acquire();
+                    final String column = columnsToGrab.get(i % columnsToGrab.size()).getName();
+                    String query = String.format("select %s from system.local", column);
+                    ResultSetFuture future = session.executeAsync(query);
+
+                    Futures.addCallback(future, new FutureCallback<ResultSet>() {
+                        @Override
+                        public void onSuccess(ResultSet result) {
+                            semaphore.release();
+                            // Expect the column that you queried to be present, if its not we got the wrong response
+                            // back.
+                            int columnIndex = result.getColumnDefinitions().getIndexOf(column);
+                            if (columnIndex == -1) {
+                                logger.error("Got response without column {}, got columns {} from Host {}.", column,
+                                        result.getColumnDefinitions(), result.getExecutionInfo().getQueriedHost());
+                                errorTrigger.countDown();
+                                return;
+                            }
+                            completed.incrementAndGet();
+                            // Block netty io thread 25% of the time.
+                            int num = random.nextInt(1);
+                            if (num == 0) {
+                                // Sleep exactly one second, should trigger retry.
+                                Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Throwable t) {
+                            semaphore.release();
+                            // NHAEs are inevitable because of low query timeouts and blocked threads.
+                            if (!(t instanceof NoHostAvailableException)) {
+                                logger.error("Unexpected error encountered.", t);
+                                errorTrigger.countDown();
+                            }
+                        }
+                    });
+                } catch (InterruptedException e) {
+                    fail("Test interrupted", e);
+                }
+                if (i % (iterations / 10) == 0) {
+                    logger.info("Submitted {} of {} requests. ({} completed successfully)", i, iterations, completed.get());
+                }
+            }
+
+            // Wait for 10 seconds for any remaining requests to possibly trigger an error, its likely
+            // that if we get to this point this will not happen.
+            Uninterruptibles.awaitUninterruptibly(errorTrigger, 10, TimeUnit.SECONDS);
+            if (errorTrigger.getCount() == 0) {
+                fail(String.format("Error triggered after %dms.", System.currentTimeMillis() - start));
+            }
+            // Sanity check to ensure that at least some requests succeeded, we expect some failures if both
+            // hosts timeout as its likely they could be blocked on the event loop.
+            assertThat(completed.get()).isGreaterThan(0);
+        } finally {
+            try {
+                // Acquire all permits to make sure all inflight requests complete.
+                if (!semaphore.tryAcquire(concurrency, 10, TimeUnit.SECONDS)) {
+                    fail("Could not acquire all permits within 10 seconds of completion.");
+                }
+            } catch (InterruptedException e) {
+                fail("Interrupted.", e);
+            }
+            cluster.close();
+        }
+    }
+}

